app/channels/event_channel.rb

```diff
@@ -1,26 +1,51 @@
 class EventChannel < ApplicationCable::Channel
   def subscribed
-    event = Event.find(params[:id])
+    @event = Event.find(params[:id])
 
-    # Unique stream for pseudo-unicast of the initial data. I hate myself for this.
-    temp_stream_name = String(event) + '|' + ('a'..'z').to_a.shuffle[0, 16].join
+    # TODO: use user identifier from cookie
+    @user_identifier = ('a'..'z').to_a.sample(16).join
+
+    temp_stream_name = String(@event) + '|' + @user_identifier
     stream_from temp_stream_name
 
-    Song.where(event: event).each do |song|
+    Song.where(event: @event).each do |song|
       ActionCable.server.broadcast(temp_stream_name, song)
     end
 
-    stream_for event
+    stream_for @event
   end
 
-  def receive(data)
-    event = Event.find(params[:id])
-
-    unless Song.exists?(uri: data['uri'], event: event)
+  def submit_song(data)
+    unless Song.exists?(uri: data['uri'], event: @event)
       song = Song.create!(name: data['name'], artist: data['artist'], art: data['art'], duration: data['duration'],
-                          uri: data['uri'], event: event)
+                          uri: data['uri'], event: @event)
+
+      EventChannel.broadcast_to(@event, song)
+    end
+  end
+
+  def vote(data)
+    # TODO: way to set vote to 0
+
+    user_identifier = @user_identifier
+    song_id = data['song']
+
+    song = Song.find_by(id: song_id, event: @event)
+
+    unless song.nil?
+      vote = Vote.find_by(user_identifier: user_identifier, song: song)
```

Comment:
@mstojcevich hey! just a note, you should be able use `song.upvote(user_identifer)` here 😺